psycopg_pool/psycopg_pool/base.py

```diff
@@ -117,6 +118,9 @@ def closed(self) -> bool:
         """`!True` if the pool is closed."""
         return self._closed
 
+    def open(self) -> None:
```

Comment:
As discussed making `open()` async is indeed desired, but it would make it hard (if ever possible) to call it in `__init__()`.
To recap, the original idea to run it as task `create_task(self.open())` is not enough because this will make the pool non-functional after object creation (thus a breaking change). Trying to wait on the task to finish (in `__ini__()`) seems also difficult: I've tried using a `threading.Event` but this does not play well with asyncio, possibly because the `Event` locks the thread in which the asyncio loop is running, for instance:
```python
import asyncio
import threading

async def unlock(ev):
    ev.set()

async def main():
    ev = threading.Event()
    asyncio.create_task(unlock(ev))
    ev.wait()  # block here
    assert ev.is_set()

if __name__ == "__main__":
    asyncio.run(main())
```
So it seems we're left with introducing a breaking change: make `open()` async and not opening the pool in `__init__()`.
I'll open another PR implementing this; then this one can be closed.

Reply:
I checked this problem: the problem is just that the thread goes to smash into the wait without a chance to run the unlock task:

adding `await asyncio.sleep(0)` before `ev.wait()` solves the problem. It doesn't even need a `run_in_executor()` - I thought it would.

So it seems we can call the moral equivalent of `await self.open()` from a sync task: as such we could implement an async open method, and opening in the aenter method, add an `open=True` parameter to the init, and pretty much release all this in 3.1, deferring to 4.0 only defaulting the open to False, if it makes sense somewhere. And of course a different class, `TrioPool`, may have whatever default it wants, to backwards compatibility to maintain.

At the moment I'm a bit lost between #151 and #155. How do you suggest to proceed?

Cheers!

Reply:
> I checked this problem: the problem is just that the thread goes to smash into the wait without a chance to run the unlock task. Adding `await asyncio.sleep(0)` before `ev.wait()` solves the problem. It doesn't even need a `run_in_executor()` - I thought it would.

Well, yes. But the case I posted is a bit too simplistic because we cannot actually emit `await asyncio.sleep(0)` in `AsyncConnectionPool.__init__()` because it's not `async`.

> So it seems we can call the moral equivalent of `await self.open()` from a sync task: as such we could implement an async open method, and opening in the aenter method, add an `open=True` parameter to the init, and pretty much release all this in 3.1, deferring to 4.0 only defaulting the open to False, if it makes sense somewhere. And of course a different class, `TrioPool`, may have whatever default it wants, no backwards compatibility to maintain.
> 
> At the moment I'm a bit lost between #151 and #155. How do you suggest to proceed?

This attempt (#151) seemed like a dead-end, so I opened #155 with a breaking change. Unless I missed something the `await asyncio.sleep(0)` would not work; so it's still a dead-end.

Alternatively, if we want to leave the dependency on `anyio` optional (the approach I think we agreed on for the "core" part of psycopg), then we'll have to implement another class `AnyIOConnnectionPool` and we can avoid breaking `AsyncConnectionPool` (at the cost of having different behaviors w.r.t. tasks startup at object initialization). That's a third way.

So:
- #151 seems like a dead-end to me, agreed?
- #155 is a breaking change with a "better" (IMHO) interface (no tasks startup at object initialization)
- but it's not strictly needed if we go with a different class for anyio (the third way, another would-be PR)